From 2d657d911905772eae4c48157b465d3145bec52f Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sun, 22 May 2022 16:09:23 +1000
Subject: [PATCH 2/3] include: Add THREAD_POWER_THROTTLING_STATE type

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 include/processthreadsapi.h | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/include/processthreadsapi.h b/include/processthreadsapi.h
index d266b7a727b..f74758d6807 100644
--- a/include/processthreadsapi.h
+++ b/include/processthreadsapi.h
@@ -37,6 +37,14 @@ typedef struct _MEMORY_PRIORITY_INFORMATION
     ULONG MemoryPriority;
 } MEMORY_PRIORITY_INFORMATION, *PMEMORY_PRIORITY_INFORMATION;
 
+#undef ControlMask
+typedef struct _THREAD_POWER_THROTTLING_STATE
+{
+    ULONG Version;
+    ULONG ControlMask;
+    ULONG StateMask;
+} THREAD_POWER_THROTTLING_STATE;
+
 WINBASEAPI HRESULT WINAPI GetThreadDescription(HANDLE,PWSTR *);
 WINBASEAPI HRESULT WINAPI SetThreadDescription(HANDLE,PCWSTR);
 WINBASEAPI BOOL WINAPI SetThreadInformation(HANDLE,THREAD_INFORMATION_CLASS,LPVOID,DWORD);
-- 
2.35.1

