From 4bf9d2403f269e7f3595ad075a4afee9adbda51f Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Wed, 6 Jul 2022 12:46:50 +0200
Subject: [PATCH] ntdll: Fix the return value of NtQueryKey for a short buffer.

Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/ntdll/unix/registry.c |  8 ++------
 2 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/dlls/ntdll/unix/registry.c b/dlls/ntdll/unix/registry.c
index 9c98bf48399f..428f8c9eb3ce 100644
--- a/dlls/ntdll/unix/registry.c
+++ b/dlls/ntdll/unix/registry.c
@@ -268,7 +268,6 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
             case KeyBasicInformation:
             {
                 KEY_BASIC_INFORMATION keyinfo;
-                fixed_size = (char *)keyinfo.Name - (char *)&keyinfo;
                 keyinfo.LastWriteTime.QuadPart = reply->modif;
                 keyinfo.TitleIndex = 0;
                 keyinfo.NameLength = reply->namelen;
@@ -279,7 +278,6 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
             case KeyFullInformation:
             {
                 KEY_FULL_INFORMATION keyinfo;
-                fixed_size = (char *)keyinfo.Class - (char *)&keyinfo;
                 keyinfo.LastWriteTime.QuadPart = reply->modif;
                 keyinfo.TitleIndex = 0;
                 keyinfo.ClassLength = wine_server_reply_size(reply);
@@ -297,7 +295,6 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
             case KeyNodeInformation:
             {
                 KEY_NODE_INFORMATION keyinfo;
-                fixed_size = (char *)keyinfo.Name - (char *)&keyinfo;
                 keyinfo.LastWriteTime.QuadPart = reply->modif;
                 keyinfo.TitleIndex = 0;
                 if (reply->namelen < wine_server_reply_size(reply))
@@ -318,7 +315,6 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
             case KeyNameInformation:
             {
                 KEY_NAME_INFORMATION keyinfo;
-                fixed_size = (char *)keyinfo.Name - (char *)&keyinfo;
                 keyinfo.NameLength = reply->namelen;
                 memcpy( info, &keyinfo, min( length, fixed_size ) );
                 break;
@@ -327,7 +323,6 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
             case KeyCachedInformation:
             {
                 KEY_CACHED_INFORMATION keyinfo;
-                fixed_size = sizeof(keyinfo);
                 keyinfo.LastWriteTime.QuadPart = reply->modif;
                 keyinfo.TitleIndex = 0;
                 keyinfo.SubKeys = reply->subkeys;
@@ -344,7 +339,8 @@ static NTSTATUS enumerate_key( HANDLE handle, int index, KEY_INFORMATION_CLASS i
                 break;
             }
             *result_len = fixed_size + reply->total;
-            if (length < *result_len) ret = STATUS_BUFFER_OVERFLOW;
+            if (length < fixed_size) ret = STATUS_BUFFER_TOO_SMALL;
+            else if (length < *result_len) ret = STATUS_BUFFER_OVERFLOW;
         }
     }
     SERVER_END_REQ;
