--- a/dlls/winmm/Makefile.in	
+++ a/dlls/winmm/Makefile.in	
@@ -1,7 +1,7 @@ 
 EXTRADEFS = -D_WINMM_
 MODULE    = winmm.dll
 IMPORTLIB = winmm
-IMPORTS   = uuid user32 advapi32 ole32 msacm32 dinput8
+IMPORTS   = uuid user32 advapi32 ole32 msacm32
 
 C_SRCS = \
 	driver.c \
--- a/dlls/winmm/joystick.c	
+++ a/dlls/winmm/joystick.c	
@@ -154,9 +154,16 @@ static BOOL CALLBACK enum_instances( const DIDEVICEINSTANCEW *instance, void *co
 
 void joystick_load( HINSTANCE instance )
 {
-    HRESULT hr = DirectInput8Create( instance, DIRECTINPUT_VERSION, &IID_IDirectInput8W,
-                                     (void **)&dinput, NULL );
-    if (FAILED(hr)) WARN( "could not create dinput instance, hr %#x\n", hr );
+    HRESULT hr;
+
+    CoInitialize( NULL );
+
+    hr = CoCreateInstance( &CLSID_DirectInput8, NULL, CLSCTX_INPROC_SERVER, &IID_IDirectInput8W, (void **)&dinput );
+    if (FAILED(hr))
+    {
+        WARN( "could not create dinput instance, hr %#x\n", hr );
+        CoUninitialize();
+    }
 }
 
 void joystick_unload()
@@ -173,6 +180,7 @@ void joystick_unload()
     }
 
     IDirectInput8_Release( dinput );
+    CoUninitialize();
 }
 
 static void find_joysticks(void)

