--- a/dlls/dxgi/dxgi_main.c	
+++ a/dlls/dxgi/dxgi_main.c
@@ -268,8 +268,6 @@ HRESULT WINAPI DXGID3D10RegisterLayers(const struct dxgi_device_layer *layers, UINT layer_count)
 
 HRESULT WINAPI DXGIGetDebugInterface1(UINT flags, REFIID iid, void **debug)
 {
-    TRACE("flags %#x, iid %s, debug %p.\n", flags, debugstr_guid(iid), debug);
-
     WARN("Returning DXGI_ERROR_SDK_COMPONENT_MISSING.\n");
     return DXGI_ERROR_SDK_COMPONENT_MISSING;
 }


