From f2347ae7216626d248cfaf9445862561b2b5eef7 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 1 Jul 2019 09:58:55 +1000
Subject: [PATCH] loader: Add Keyboard Layouts registry enteries.

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 loader/wine.inf.in | 209 +++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 209 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index ff601e41b26..6ffb8c56d87 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -63,6 +63,7 @@ AddReg=\
     Debugger,\
     DirectX,\
     Fonts,\
+    KeyboardLayouts,\
     MCI,\
     Misc,\
     OLE,\
@@ -87,6 +88,7 @@ AddReg=\
     Debugger,\
     DirectX,\
     Fonts,\
+    KeyboardLayouts,\
     MCI,\
     Misc,\
     OLE,\
@@ -113,6 +115,7 @@ AddReg=\
     Debugger,\
     DirectX,\
     Fonts,\
+    KeyboardLayouts,\
     MCI,\
     Misc,\
     OLE,\
@@ -160,6 +163,7 @@ AddReg=\
     CurrentVersionWow64,\
     Debugger,\
     DirectX,\
+    KeyboardLayouts,\
     MCI,\
     Misc,\
     Tapi,\
@@ -597,6 +601,211 @@ HKCU,Software\Microsoft\Windows\Shell\Associations\UrlAssociations\ftp\UserChoic
 HKCU,Software\Microsoft\Windows\Shell\Associations\UrlAssociations\http\UserChoice,"ProgId",,"http"
 HKCU,Software\Microsoft\Windows\Shell\Associations\UrlAssociations\https\UserChoice,"ProgId",,"https"
 
+[KeyboardLayouts]
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000401,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000402,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000404,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000405,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000406,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000407,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000040f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000410,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000411,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000412,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000413,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000414,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000415,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000416,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000418,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000419,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000041f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000420,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000422,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000423,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000424,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000425,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000426,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000427,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000428,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000429,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000042a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000042b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000042c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000042e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000042f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000432,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000437,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000438,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000439,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000043a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000043b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000043f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000440,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000442,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000444,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000445,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000446,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000447,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000448,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000449,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000044a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000044b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000044c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000044d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000044e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000450,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000451,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000452,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000453,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000454,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000045a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000045b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000045c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000461,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000463,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000465,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000468,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000046a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000046c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000046d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000046e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000046f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000470,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000474,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000475,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000480,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000481,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000485,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000488,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000492,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000804,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000807,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000809,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000080a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000080c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000813,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000816,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000081a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000082c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000083b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000843,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000850,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000085d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000085f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000c04,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000c0c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000c1a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00001004,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00001009,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000100c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000105f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00001404,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00001809,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0000201a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00004009,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010401,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010402,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010405,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010407,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001040a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001040e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010410,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010415,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010416,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010418,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010419,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001041b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001041e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001041f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010426,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010427,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001042b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001042c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001042e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001042f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010437,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010439,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001043a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001043b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010444,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010445,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010451,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010453,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001045a,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001045b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001045c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001045d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010465,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010480,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001080c,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001083b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010850,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00010c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00011009,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0001105f,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00011809,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020401,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020402,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020405,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0002040d,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020418,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020419,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0002041e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020422,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020426,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020427,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0002042b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0002042e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020437,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020445,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0002083b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00020c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00030402,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00030408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00030409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0003041e,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\0003042b,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00030437,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00030c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00040402,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00040408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00040409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00040437,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00040c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00050408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00050409,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00050429,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00060408,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00070c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00080c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00090c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000a0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000b0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000c0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000d0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000e0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\000f0c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00100c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00110c00,,16
+HKLM,SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00120c00,,16
+
 [OLE]
 HKLM,"Software\Microsoft\OLE","EnableDCOM",,"Y"
 HKLM,"Software\Microsoft\OLE","EnableRemoteConnect",,"N"
-- 
2.35.1

