From 71be762c3d84b84c3d9249f80467797ce3519bb6 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 12 Sep 2020 16:31:09 +1000
Subject: Add support for private contexts


diff --git a/libs/faudio/include/FACT.h b/libs/faudio/include/FACT.h
index 579b7168812..084b1e55cfe 100644
--- a/libs/faudio/include/FACT.h
+++ b/libs/faudio/include/FACT.h
@@ -805,6 +805,18 @@ FACTAPI uint32_t FACTCue_SetOutputVoiceMatrix(
 	const float *pLevelMatrix /* SourceChannels * DestinationChannels */
 );
 
+FACTAPI void FACTWave_SetPrivateContext(FACTWave *pWave, void *context);
+FACTAPI void* FACTWave_GetPrivateContext(FACTWave *pWave);
+
+FACTAPI void FACTWaveBank_SetPrivateContext(FACTWaveBank *pWaveBank, void *context);
+FACTAPI void* FACTWaveBank_GetPrivateContext(FACTWaveBank *pWaveBank);
+
+FACTAPI void FACTSoundBank_SetPrivateContext(FACTSoundBank *pSoundBank, void *context);
+FACTAPI void* FACTSoundBank_GetPrivateContext(FACTSoundBank *pSoundBank);
+
+FACTAPI void FACTCue_SetPrivateContext(FACTCue *pCue, void *context);
+FACTAPI void* FACTCue_GetPrivateContext(FACTCue *pCue);
+
 #ifdef __cplusplus
 }
 #endif /* __cplusplus */
diff --git a/libs/faudio/src/FACT.c b/libs/faudio/src/FACT.c
index 5eca83b389f..e2e7b10058d 100644
--- a/libs/faudio/src/FACT.c
+++ b/libs/faudio/src/FACT.c
@@ -1177,6 +1177,9 @@ uint32_t FACTSoundBank_Prepare(
 	(*ppCue)->notifyOnDestroy = 0;
 	(*ppCue)->usercontext = NULL;
 
+	/* User data */
+	(*ppCue)->privatecontext = NULL;
+
 	/* Sound data */
 	(*ppCue)->data = &pSoundBank->cues[nCueIndex];
 	if ((*ppCue)->data->flags & 0x04)
@@ -1798,6 +1801,9 @@ uint32_t FACTWaveBank_Prepare(
 	(*ppWave)->pitch = 0;
 	(*ppWave)->loopCount = nLoopCount;
 
+	/* User data */
+	(*ppWave)->privatecontext = NULL;
+
 	/* TODO: Convert dwPlayOffset to a byte offset */
 	FAudio_assert(dwPlayOffset == 0);
 #if 0
@@ -2175,11 +2181,14 @@ uint32_t FACTWave_Stop(FACTWave *pWave, uint32_t dwFlags)
 	{
 		FACTNotification note;
 		note.type = FACTNOTIFICATIONTYPE_WAVESTOP;
+		note.wave.cueIndex = pWave->parentCue->index;
+		note.wave.pCue = pWave->parentCue;
+		note.wave.pSoundBank = pWave->parentCue->parentBank;
 		note.wave.pWave = pWave;
-		if (pWave->parentBank->parentEngine->notifications & NOTIFY_WAVESTOP)
-		{
-			note.pvContext = pWave->parentBank->parentEngine->wave_context;
-		}
+		note.wave.pWaveBank = pWave->parentBank;
+
+		note.pvContext = pWave->parentBank->parentEngine->wave_context;
+
 		pWave->parentBank->parentEngine->notificationCallback(&note);
 	}
 
@@ -3018,4 +3027,42 @@ uint32_t FACTCue_SetOutputVoiceMatrix(
 	return 0;
 }
 
+void FACTWave_SetPrivateContext(FACTWave *pWave, void *context)
+{
+    pWave->privatecontext = context;
+}
+void* FACTWave_GetPrivateContext(FACTWave *pWave)
+{
+    return pWave->privatecontext;
+}
+
+FACTAPI void FACTWaveBank_SetPrivateContext(FACTWaveBank *pWaveBank, void *context)
+{
+    pWaveBank->privatecontext = context;
+}
+
+FACTAPI void* FACTWaveBank_GetPrivateContext(FACTWaveBank *pWaveBank)
+{
+    return pWaveBank->privatecontext;
+}
+
+FACTAPI void FACTSoundBank_SetPrivateContext(FACTSoundBank *pSoundBank, void *context)
+{
+    pSoundBank->privatecontext = context;
+}
+FACTAPI void* FACTSoundBank_GetPrivateContext(FACTSoundBank *pSoundBank)
+{
+    return pSoundBank->privatecontext;
+}
+
+FACTAPI void FACTCue_SetPrivateContext(FACTCue *pCue, void *context)
+{
+    pCue->privatecontext = context;
+}
+
+FACTAPI void* FACTCue_GetPrivateContext(FACTCue *pCue)
+{
+    return pCue->privatecontext;
+}
+
 /* vim: set noexpandtab shiftwidth=8 tabstop=8: */
diff --git a/libs/faudio/src/FACT_internal.h b/libs/faudio/src/FACT_internal.h
index 6bf522e64c4..52d07e5f9d3 100644
--- a/libs/faudio/src/FACT_internal.h
+++ b/libs/faudio/src/FACT_internal.h
@@ -473,6 +473,9 @@ struct FACTSoundBank
 	uint32_t *variationCodes;
 	FACTTransitionTable *transitions;
 	uint32_t *transitionCodes;
+
+	/* Application data */
+	void *privatecontext;
 };
 
 struct FACTWaveBank
@@ -498,6 +501,9 @@ struct FACTWaveBank
 	uint8_t *packetBuffer;
 	uint32_t packetBufferLen;
 	void* io;
+
+	/* Application data */
+	void *privatecontext;
 };
 
 struct FACTWave
@@ -524,6 +530,9 @@ struct FACTWave
 	uint16_t srcChannels;
 	FAudioSourceVoice *voice;
 	FACTWaveCallback callback;
+
+	/* Application data */
+	void *privatecontext;
 };
 
 struct FACTCue
@@ -569,6 +578,9 @@ struct FACTCue
 	/* Timer */
 	uint32_t start;
 	uint32_t elapsed;
+
+	/* Application data */
+	void *privatecontext;
 };
 
 /* Internal functions */
