From a8f9265b71b423e2d280645a225465750a1d4594 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 25 Oct 2021 09:48:50 +1100
Subject: [PATCH 3/3] Send NOTIFY_CUESTOP when Stop is called

---
 libs/faudio/src/FACT.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libs/faudio/src/FACT.c b/libs/faudio/src/FACT.c
index cea960d6710..d9c5d7c2766 100644
--- a/libs/faudio/src/FACT.c
+++ b/libs/faudio/src/FACT.c
@@ -2685,6 +2685,8 @@ uint32_t FACTCue_Stop(FACTCue *pCue, uint32_t dwFlags)
 		}
 	}
 
+	FACT_INTERNAL_SendCueNotification(pCue, NOTIFY_CUESTOP, FACTNOTIFICATIONTYPE_CUESTOP);
+
 	FAudio_PlatformUnlockMutex(pCue->parentBank->parentEngine->apiLock);
 	return 0;
 }
-- 
2.33.0

