diff --git a/dlls/dinput/joystick_sdl.c b/dlls/dinput/joystick_sdl.c
index cd26e1b2df..fcba4aaada 100644
--- a/dlls/dinput/joystick_sdl.c
+++ b/dlls/dinput/joystick_sdl.c
@@ -491,6 +491,11 @@ static JoystickImpl *alloc_device(REFGUID rguid, IDirectInputImpl *dinput, unsig
     list_add_tail(&dinput->devices_list, &newDevice->generic.base.entry);
     LeaveCriticalSection(&dinput->crit);
 
+    newDevice->generic.base.buffersize = 20;
+    newDevice->generic.base.queue_len = 20;
+    newDevice->generic.base.data_queue = HeapAlloc(GetProcessHeap(), 0,
+                newDevice->generic.base.queue_len * sizeof(DIDEVICEOBJECTDATA));
+
     return newDevice;
 
 failed:
