From 82c64b9c951777b0143e3125f4b95a3ac6e091c3 Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Fri, 2 Oct 2020 14:01:13 +0300
Subject: [PATCH] evr/presenter: Add missing method stubs.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/evr/presenter.c | 48 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)

diff --git a/dlls/evr/presenter.c b/dlls/evr/presenter.c
index 0663383b14f..04e9a556840 100644
--- a/dlls/evr/presenter.c
+++ b/dlls/evr/presenter.c
@@ -563,6 +563,48 @@ static HRESULT WINAPI video_presenter_control_GetCurrentImage(IMFVideoDisplayCon
     return E_NOTIMPL;
 }
 
+static HRESULT WINAPI video_presenter_control_SetBorderColor(IMFVideoDisplayControl *iface, COLORREF color)
+{
+    FIXME("%p, %#x.\n", iface, color);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI video_presenter_control_GetBorderColor(IMFVideoDisplayControl *iface, COLORREF *color)
+{
+    FIXME("%p, %p.\n", iface, color);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI video_presenter_control_SetRenderingPrefs(IMFVideoDisplayControl *iface, DWORD flags)
+{
+    FIXME("%p, %#x.\n", iface, flags);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI video_presenter_control_GetRenderingPrefs(IMFVideoDisplayControl *iface, DWORD *flags)
+{
+    FIXME("%p, %p.\n", iface, flags);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI video_presenter_control_SetFullscreen(IMFVideoDisplayControl *iface, BOOL fullscreen)
+{
+    FIXME("%p, %d.\n", iface, fullscreen);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI video_presenter_control_GetFullscreen(IMFVideoDisplayControl *iface, BOOL *fullscreen)
+{
+    FIXME("%p, %p.\n", iface, fullscreen);
+
+    return E_NOTIMPL;
+}
+
 static const IMFVideoDisplayControlVtbl video_presenter_control_vtbl =
 {
     video_presenter_control_QueryInterface,
@@ -578,6 +620,12 @@ static const IMFVideoDisplayControlVtbl video_presenter_control_vtbl =
     video_presenter_control_GetVideoWindow,
     video_presenter_control_RepaintVideo,
     video_presenter_control_GetCurrentImage,
+    video_presenter_control_SetBorderColor,
+    video_presenter_control_GetBorderColor,
+    video_presenter_control_SetRenderingPrefs,
+    video_presenter_control_GetRenderingPrefs,
+    video_presenter_control_SetFullscreen,
+    video_presenter_control_GetFullscreen,
 };
 
 static HRESULT WINAPI video_presenter_rate_support_QueryInterface(IMFRateSupport *iface, REFIID riid, void **obj)
