From ae2a2bdbe8b35cd7b23c1049a4c1615a4ce69a9f Mon Sep 17 00:00:00 2001
From: Hans-Kristian Arntzen <post@arntzen-software.no>
Date: Thu, 5 Mar 2020 15:36:05 +0100
Subject: [PATCH] vkd3d: Fix DXIL check in test_coverage.

Signed-off-by: Hans-Kristian Arntzen <post@arntzen-software.no>
---
 tests/d3d12.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/tests/d3d12.c b/tests/d3d12.c
index f1e0122c..92f33ca7 100644
--- a/tests/d3d12.c
+++ b/tests/d3d12.c
@@ -33736,6 +33736,13 @@ static void test_coverage(bool use_dxil)
     desc.no_root_signature = true;
     if (!init_test_context(&context, &desc))
         return;
+
+    if (use_dxil && !context_supports_dxil(&context))
+    {
+        destroy_test_context(&context);
+        return;
+    }
+
     command_list = context.list;
     queue = context.queue;
 
