diff --git a/libs/vkd3d/command.c b/libs/vkd3d/command.c
index c401349..62b4751 100644
--- a/libs/vkd3d/command.c
+++ b/libs/vkd3d/command.c
@@ -1862,14 +1862,17 @@ static void d3d12_command_list_invalidate_current_render_pass(struct d3d12_comma
 static void d3d12_command_list_invalidate_bindings(struct d3d12_command_list *list,
         struct d3d12_pipeline_state *state)
 {
+    struct vkd3d_pipeline_bindings *bindings;
+
     if (!state)
         return;
 
+    bindings = &list->pipeline_bindings[state->vk_bind_point];
+
+    bindings->descriptor_table_dirty_mask |= bindings->descriptor_table_active_mask;
+
     if (state->uav_counter_mask)
-    {
-        struct vkd3d_pipeline_bindings *bindings = &list->pipeline_bindings[state->vk_bind_point];
         bindings->uav_counter_dirty_mask = ~(uint8_t)0;
-    }
 }
 
 static void d3d12_command_list_invalidate_root_parameters(struct d3d12_command_list *list,
