diff --git a/src/util/config/config.cpp b/src/util/config/config.cpp
index da68b59a..b560663a 100644
--- a/src/util/config/config.cpp
+++ b/src/util/config/config.cpp
@@ -423,6 +423,14 @@ namespace dxvk {
     { R"(\\k2\.exe$)", {{
       { "d3d9.memoryTrackTest",             "True" },
     }} },
+    /* Ninja Gaiden Sigma                        */
+    { R"(\\NINJA GAIDEN SIGMA\.exe$)", {{
+      { "d3d9.deferSurfaceCreation",        "True" },
+    }} },
+    /* Ninja Gaiden Sigma 2                      */
+    { R"(\\NINJA GAIDEN SIGMA2\.exe$)", {{
+      { "d3d9.deferSurfaceCreation",        "True" },
+    }} },
   }};
 
 
