diff --git a/src/util/config/config.cpp b/src/util/config/config.cpp
index f3aaed0f..45a71721 100644
--- a/src/util/config/config.cpp
+++ b/src/util/config/config.cpp
@@ -681,9 +681,14 @@ namespace dxvk {
     { R"(\\bms\.exe$)", {{
       { "d3d9.customVendorId",              "10de" },
     }} },
+    /* Secret World: Legends                   *
+     * Launcher needs shaderModel 2            *
+     * to not crash                            */
+    { R"(\\ClientPatcher\.exe$)", {{
+      { "d3d9.shaderModel",              "2" },
+    }} },
   }};
 
-
   static bool isWhitespace(char ch) {
     return ch == ' ' || ch == '\x9' || ch == '\r';
   }
