diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 368448c9f8d..a8169cff07b 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -2633,6 +2633,7 @@ static NTSTATUS load_dll( const WCHAR *load_path, const WCHAR *libname, const WC
     void *module;
     SECTION_IMAGE_INFORMATION image_info;
     NTSTATUS nts;
+    BOOL so_dll = FALSE;
 
     TRACE( "looking for %s in %s\n", debugstr_w(libname), debugstr_w(load_path) );
 
@@ -2665,6 +2666,7 @@ static NTSTATUS load_dll( const WCHAR *load_path, const WCHAR *libname, const WC
         case LO_BUILTIN_NATIVE:
         case LO_DEFAULT:
             if (!load_so_dll( load_path, &nt_name, flags, pwm )) nts = STATUS_SUCCESS;
+            so_dll = TRUE;
             break;
         default:
             nts = STATUS_DLL_NOT_FOUND;
@@ -2748,7 +2750,33 @@ static NTSTATUS load_dll( const WCHAR *load_path, const WCHAR *libname, const WC
 
 done:
     if (nts == STATUS_SUCCESS)
+    {
         TRACE("Loaded module %s at %p\n", debugstr_us(&nt_name), (*pwm)->ldr.DllBase);
+
+        /* hack: ensure timedatestamp != 0 */
+        if (!so_dll)
+        {
+            ULONG old_prot, temp;
+            SIZE_T len;
+            void *base = (*pwm)->ldr.DllBase;
+
+            len = PAGE_SIZE;
+
+            if (!(nts = NtProtectVirtualMemory(NtCurrentProcess(), &base, &len, PAGE_EXECUTE_READWRITE, &old_prot)))
+            {
+                PIMAGE_NT_HEADERS headers = RtlImageNtHeader(base);
+                if (headers->FileHeader.TimeDateStamp == 0)
+                    headers->FileHeader.TimeDateStamp = 0xdeadbeef;
+
+                nts = NtProtectVirtualMemory(NtCurrentProcess(), &base, &len, old_prot, &temp);
+            }
+            if (nts != STATUS_SUCCESS)
+            {
+                ERR("Failed to modify timedatestamp, err %#x\n", nts);
+                nts = STATUS_SUCCESS;
+            }
+        }
+    }
     else
         WARN("Failed to load module %s; status=%x\n", debugstr_w(libname), nts);
 
diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
index c2b6ea603e3..b65a32b47b0 100644
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -1204,7 +1204,29 @@ static NTSTATUS open_dll_file( const char *name, void **module, SECTION_IMAGE_IN
         status = STATUS_IMAGE_MACHINE_TYPE_MISMATCH;
     }
 
-    if (!status) status = add_builtin_module( *module, NULL, &st );
+    if (!status)
+    {
+        ULONG old_prot, temp;
+        SIZE_T len;
+
+        status = add_builtin_module( *module, NULL, &st );
+
+        /* hack: ensure timedatestamp != 0 */
+        len = 0x1000;
+        if (!(status = NtProtectVirtualMemory(NtCurrentProcess(), module, &len, PAGE_EXECUTE_READWRITE, &old_prot)))
+        {
+            PIMAGE_NT_HEADERS headers = get_rva( *module, ((IMAGE_DOS_HEADER *)*module)->e_lfanew );
+            if (headers->FileHeader.TimeDateStamp == 0)
+                headers->FileHeader.TimeDateStamp = 0xdeadbeef;
+
+            status = NtProtectVirtualMemory(NtCurrentProcess(), module, &len, old_prot, &temp);
+        }
+        if (status != STATUS_SUCCESS)
+        {
+            ERR("Failed to modify timedatestamp, err %#x\n", status);
+            status = STATUS_SUCCESS;
+        }
+    }
 
     if (status)
     {
