diff --git a/dlls/wintrust/wintrust_main.c b/dlls/wintrust/wintrust_main.c
index 524f4a4d5a..afc6e81bca 100644
--- a/dlls/wintrust/wintrust_main.c
+++ b/dlls/wintrust/wintrust_main.c
@@ -304,6 +304,12 @@ static LONG WINTRUST_DefaultVerify(HWND hwnd, GUID *actionID,
     numSteps = WINTRUST_AddTrustStepsFromFunctions(verifySteps,
      provData->psPfns);
     err = WINTRUST_ExecuteSteps(verifySteps, numSteps, provData);
+
+    const char *sgi = getenv("SteamGameId");
+    if (sgi && !strcmp(sgi, "637650")) {
+        err = S_OK;
+    }
+
     goto done;
 
 error:
