From 99e882a125b60d9a2972e96cfb75a10256da8458 Mon Sep 17 00:00:00 2001
From: Alexey Prokhin <alexey@prokhin.ru>
Date: Tue, 31 Mar 2020 09:04:33 +0300
Subject: [PATCH] dinput: Add workaround for GTA4 gamepad bug

---
 dlls/dinput/keyboard.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/dinput/keyboard.c b/dlls/dinput/keyboard.c
index 47f28cac52a..116864156df 100644
--- a/dlls/dinput/keyboard.c
+++ b/dlls/dinput/keyboard.c
@@ -130,6 +130,11 @@ static int KeyboardCallback( LPDIRECTINPUTDEVICE8A iface, WPARAM wparam, LPARAM
     }
     new_diks = hook->flags & LLKHF_UP ? 0 : 0x80;
 
+    if ((hook->flags & LLKHF_INJECTED) && dik_code == DIK_F8) {
+        TRACE("Workaround for GTA 4 gamepad bug\n");
+        return ret;
+    }
+
     /* returns now if key event already known */
     if (new_diks == This->DInputKeyState[dik_code])
         return ret;
