diff --git a/dlls/ntdll/path.c b/dlls/ntdll/path.c
index 32e1f951af..b3cd74ecf7 100644
--- a/dlls/ntdll/path.c
+++ b/dlls/ntdll/path.c
@@ -545,6 +545,7 @@ static ULONG get_full_path_helper(LPCWSTR name, LPWSTR buffer, ULONG size)
     LPCWSTR                     ptr;
     const UNICODE_STRING*       cd;
     WCHAR                       tmp[4];
+    const char *sgi = getenv("SteamGameId");
 
     /* return error if name only consists of spaces */
     for (ptr = name; *ptr; ptr++) if (*ptr != ' ') break;
@@ -552,10 +553,13 @@ static ULONG get_full_path_helper(LPCWSTR name, LPWSTR buffer, ULONG size)
 
     RtlAcquirePebLock();
 
-    if (NtCurrentTeb()->Tib.SubSystemTib)  /* FIXME: hack */
-        cd = &((WIN16_SUBSYSTEM_TIB *)NtCurrentTeb()->Tib.SubSystemTib)->curdir.DosPath;
-    else
-        cd = &NtCurrentTeb()->Peb->ProcessParameters->CurrentDirectory.DosPath;
+    cd = &NtCurrentTeb()->Peb->ProcessParameters->CurrentDirectory.DosPath;
+    
+    if (NtCurrentTeb()->Tib.SubSystemTib) {      /* FIXME: hack */
+        if ((!sgi) | (sgi && strcmp(sgi, "342200"))) {
+            cd = &((WIN16_SUBSYSTEM_TIB *)NtCurrentTeb()->Tib.SubSystemTib)->curdir.DosPath;
+        }
+    }
 
     switch (RtlDetermineDosPathNameType_U(name))
     {
