From 344386726f06cc3507d99eeea580b78f2bb02ad4 Mon Sep 17 00:00:00 2001
From: udf <tabhooked@gmail.com>
Date: Mon, 27 Jun 2022 01:14:45 +0200
Subject: [PATCH] file: add WINE_NO_OPEN_FILE_SEARCH env var

---
 dlls/ntdll/unix/file.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 850c70a6b2b..67b134998b2 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -3188,6 +3188,17 @@ static NTSTATUS lookup_unix_name( const WCHAR *name, int name_len, char **buffer
     if (is_unix && (disposition == FILE_OPEN || disposition == FILE_OVERWRITE))
         return STATUS_OBJECT_NAME_NOT_FOUND;
 
+    static int skip_search = -1;
+    if (skip_search == -1)
+    {
+        const char *env_var;
+
+        if ((skip_search = (env_var = getenv("WINE_NO_OPEN_FILE_SEARCH")) && atoi(env_var)))
+            WARN("Disabling case insensitive search for opening files");
+    }
+    if (skip_search && disposition == FILE_OPEN)
+        return STATUS_OBJECT_NAME_NOT_FOUND;
+
     /* now do it component by component */
 
     while (name_len)
