--- a/dlls/wined3d/adapter_gl.c	
+++ a/dlls/wined3d/adapter_gl.c	
@@ -3928,6 +3928,7 @@ static BOOL wined3d_adapter_init_gl_caps(struct wined3d_adapter *adapter,
     }
     fixup_extensions(gl_info, caps_gl_ctx, gl_renderer_str, gl_vendor,
             gpu_description->vendor, gpu_description->device);
+    d3d_info->clip_control = !!gl_info->supported[ARB_CLIP_CONTROL];
     wined3d_driver_info_init(driver_info, gpu_description, vram_bytes, 0);
     TRACE("Reporting (fake) driver version 0x%08x-0x%08x.\n",
             driver_info->version_high, driver_info->version_low);
