From 5481c579e00679173f7c8d77f6701503ae878c8e Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Tue, 12 May 2020 17:05:41 -0500
Subject: [PATCH 16/53] winegstreamer: Translate WMA caps to attributes.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/winegstreamer/mfplat.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/dlls/winegstreamer/mfplat.c b/dlls/winegstreamer/mfplat.c
index c45cfacc08..fe187e6cd1 100644
--- a/dlls/winegstreamer/mfplat.c
+++ b/dlls/winegstreamer/mfplat.c
@@ -831,6 +831,30 @@ static IMFMediaType* transform_to_media_type(GstCaps *caps)
                     FIXME("Unhandled mpegversion %d\n", mpeg_version);
             }
         }
+        else if (!(strcmp(mime_type, "audio/x-wma")))
+        {
+            gint wma_version, block_align;
+
+            if (gst_structure_get_int(info, "wmaversion", &wma_version))
+            {
+                switch (wma_version)
+                {
+                    case 2:
+                        IMFMediaType_SetGUID(media_type, &MF_MT_SUBTYPE, &MFAudioFormat_WMAudioV8);
+                        break;
+                    case 3:
+                        IMFMediaType_SetGUID(media_type, &MF_MT_SUBTYPE, &MFAudioFormat_WMAudioV9);
+                        break;
+                    default:
+                        FIXME("Unrecognized wmaversion %d\n", wma_version);
+                }
+            }
+
+            if (gst_structure_get_int(info, "block_align", &block_align))
+                IMFMediaType_SetUINT32(media_type, &MF_MT_AUDIO_BLOCK_ALIGNMENT, block_align);
+
+            codec_data_to_user_data(info, media_type);
+        }
         else
             FIXME("Unrecognized audio format %s\n", mime_type);
     }
-- 
2.26.2

