diff --git a/dlls/user32/focus.c b/dlls/user32/focus.c
index fb4d9e69fb..dd239faed8 100644
--- a/dlls/user32/focus.c
+++ b/dlls/user32/focus.c
@@ -169,6 +169,9 @@ static BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus )
         if (GetAncestor( hwnd, GA_PARENT ) == GetDesktopWindow())
             PostMessageW( GetDesktopWindow(), WM_PARENTNOTIFY, WM_NCACTIVATE, (LPARAM)hwnd );
 
+        if (hwnd == GetForegroundWindow() && !IsIconic( hwnd ))
+            USER_Driver->pSetActiveWindow( hwnd );
+
         imm_activate_window( hwnd );
     }
 
