From 8bfb9dac9e82749366fd0c441eb22eb4ef9a02d8 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Fri, 23 Oct 2020 18:23:09 +0300
Subject: [PATCH] lsteamclient: Fix end of string check in relative path
 conversion.

---
 steamclient_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/steamclient_main.c b/steamclient_main.c
index 9ff979a..9babce0 100644
--- a/steamclient_main.c
+++ b/steamclient_main.c
@@ -164,7 +164,7 @@ bool steamclient_dos_path_to_unix_path(const char *src, char *dst, int is_url)
         const char *s;
         char *d;
 
-        for(s = src, d = dst; *src; ++s, ++d){
+        for(s = src, d = dst; *s; ++s, ++d){
             if(*s == '\\')
                 *d = '/';
             else
-- 
2.26.2

