From 1301b70733c9d5256b5f8352e9bbd6a3578ba590 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 18 Mar 2021 16:54:44 -0400
Subject: [PATCH 35/39] mfplat: Stub out MFCreateDXGIDeviceManager, to avoid
 the d3d path.

---
 dlls/mfplat/main.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index dbfaeeac447..0a2f24fab8a 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -8905,10 +8905,14 @@ static const IMFDXGIDeviceManagerVtbl dxgi_device_manager_vtbl =
 HRESULT WINAPI MFCreateDXGIDeviceManager(UINT *token, IMFDXGIDeviceManager **manager)
 {
     struct dxgi_device_manager *object;
+    const char *sgi = getenv("SteamGameId");

     TRACE("%p, %p.\n", token, manager);

-    return E_NOTIMPL;
+    if (sgi && !(strcmp(sgi, "1020540") == 0 || strcmp(sgi, "1113560") == 0 || strcmp(sgi, "1449280") == 0 || strcmp(sgi, "1240440") == 0))
+    {
+        return E_NOTIMPL;
+    }

     if (!token || !manager)
         return E_POINTER;
--
2.30.2

