From 1301b70733c9d5256b5f8352e9bbd6a3578ba590 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 18 Mar 2021 16:54:44 -0400
Subject: [PATCH 35/39] mfplat: Stub out MFCreateDXGIDeviceManager, to avoid
 the d3d path.

---
 dlls/mfplat/main.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index 5406ceb5063..764d825a102 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -8881,27 +8881,33 @@ static const IMFDXGIDeviceManagerVtbl dxgi_device_manager_vtbl =
 HRESULT WINAPI MFCreateDXGIDeviceManager(UINT *token, IMFDXGIDeviceManager **manager)
 {
     struct dxgi_device_manager *object;
+    const char *sgi = getenv("SteamGameId");

     TRACE("%p, %p.\n", token, manager);

-    if (!token || !manager)
-        return E_POINTER;
+    if (sgi && (!strcmp(sgi,"1113560")))
+    {
+        if (!token || !manager)
+            return E_POINTER;

-    if (!(object = calloc(1, sizeof(*object))))
-        return E_OUTOFMEMORY;
+        if (!(object = calloc(1, sizeof(*object))))
+            return E_OUTOFMEMORY;

-    object->IMFDXGIDeviceManager_iface.lpVtbl = &dxgi_device_manager_vtbl;
-    object->refcount = 1;
-    object->token = GetTickCount();
-    InitializeCriticalSection(&object->cs);
-    InitializeConditionVariable(&object->lock);
+        object->IMFDXGIDeviceManager_iface.lpVtbl = &dxgi_device_manager_vtbl;
+        object->refcount = 1;
+        object->token = GetTickCount();
+        InitializeCriticalSection(&object->cs);
+        InitializeConditionVariable(&object->lock);

-    TRACE("Created device manager: %p, token: %u.\n", object, object->token);
+        TRACE("Created device manager: %p, token: %u.\n", object, object->token);

-    *token = object->token;
-    *manager = &object->IMFDXGIDeviceManager_iface;
+        *token = object->token;
+        *manager = &object->IMFDXGIDeviceManager_iface;

-    return S_OK;
+        return S_OK;
+    } else {
+        return E_NOTIMPL;
+    }
 }

 /***********************************************************************
--
2.30.2

