diff --git a/patches/patchinstall.sh b/patches/patchinstall.sh
index 97b65fafd..44061a2c1 100755
--- a/patches/patchinstall.sh
+++ b/patches/patchinstall.sh
@@ -3070,13 +3070,17 @@ fi
 # | 	should run unelevated by default with Vista+ setting)
 # |
 # | Modified files:
-# |   *	dlls/msi/custom.c, dlls/shell32/shlexec.c, loader/wine.inf.in, server/process.c
+# |   *	dlls/kernelbase/process.c, dlls/msi/custom.c, dlls/ntdll/process.c, dlls/ntdll/unix/env.c, dlls/shell32/shlexec.c,
+# | 	loader/wine.inf.in, server/process.c
 # |
 if test "$enable_server_default_integrity" -eq 1; then
 	patch_apply server-default_integrity/0001-server-Create-processes-using-a-limited-administrato.patch
 	patch_apply server-default_integrity/0002-shell32-Implement-the-runas-verb.patch
 	patch_apply server-default_integrity/0003-wine.inf-Set-the-EnableLUA-value-to-1.patch
 	patch_apply server-default_integrity/0004-msi-Create-the-custom-action-server-as-an-elevated-p.patch
+	patch_apply server-default_integrity/0005-ntdll-Always-start-the-initial-process-through-start.patch
+	patch_apply server-default_integrity/0006-kernelbase-Elevate-processes-if-requested-in-CreateP.patch
+	patch_apply server-default_integrity/0007-ntdll-Elevate-processes-if-requested-in-RtlCreateUse.patch
 fi
 
 # Patchset setupapi-DiskSpaceList
