From: Tk-Glitch <ti3nou@gmail.com>
Date: Wed, 27 Mar 2019 03:20:02 +0100
Subject: Add no_kernel_obj_list object method to esync. (4.5+)


diff --git a/server/esync.c b/server/esync.c
index a0745a2306..b712867e4b 100644
--- a/server/esync.c
+++ b/server/esync.c
@@ -136,6 +136,7 @@ const struct object_ops esync_ops =
     directory_link_name,       /* link_name */
     default_unlink_name,       /* unlink_name */
     no_open_file,              /* open_file */
+    no_kernel_obj_list,        /* get_kernel_obj_list */
     no_close_handle,           /* close_handle */
     esync_destroy              /* destroy */
 };
