From d7ff3995662debb178e853626d292a103a42b7d6 Mon Sep 17 00:00:00 2001
From: Steven Noonan <steven@valvesoftware.com>
Date: Tue, 23 Oct 2018 15:30:57 -0700
Subject: [PATCH] wine.inf: apply LargeAddressAware fix for Bayonetta

Signed-off-by: Steven Noonan <steven@valvesoftware.com>
---
 loader/wine.inf.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 977c8ffbc64..08e4197c55c 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -3713,6 +3713,7 @@ HKLM,Software\Wine\LicenseInformation,"Shell-PremiumInBoxGames-Chess-EnableGame"
 HKCU,Software\Valve\Steam\ActiveProcess,"PID",0x10001,0x0000fffe
 HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll",,"%16422%\Steam\steamclient.dll"
 HKCU,Software\Valve\Steam\ActiveProcess,"SteamPath",,"%16422%\Steam"
+HKCU,Software\Wine\AppDefaults\Bayonetta.exe,"LargeAddressAware",0x10003,1
 ;;Likely want *80 and *90 too, but those require removing Wine's manifest files.
 HKCU,Software\Wine\DllOverrides,"atl100",0x2,"native,builtin"
 HKCU,Software\Wine\DllOverrides,"msvcp100",0x2,"native,builtin"
@@ -3729,6 +3729,7 @@ HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll",,"%16426%\Steam\steamcl
 HKCU,Software\Valve\Steam\ActiveProcess,"SteamClientDll64",,"%16426%\Steam\steamclient64.dll"
 HKCU,Software\Valve\Steam\ActiveProcess,"SteamPath",,"%16426%\Steam"
 HKLM,Software\Wow6432Node\Valve\Steam,"InstallPath",,"%16422%\Steam"
+HKCU,Software\Wine\AppDefaults\Bayonetta.exe,"LargeAddressAware",0x10003,1
 ;;Likely want *80 and *90 too, but those require removing Wine's manifest files.
 HKCU,Software\Wine\DllOverrides,"atl100",0x2,"native,builtin"
 HKCU,Software\Wine\DllOverrides,"msvcp100",0x2,"native,builtin"

