From 206919099680fe885f0e7b7644c33e0bcfef04f3 Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Wed, 8 Aug 2018 08:38:34 -0500
Subject: [PATCH] Avoid spamming log with requests for IWineD3D11Texture2D

---
 src/d3d11/d3d11_texture.cpp | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/d3d11/d3d11_texture.cpp b/src/d3d11/d3d11_texture.cpp
index c0687382..71a2a2e4 100644
--- a/src/d3d11/d3d11_texture.cpp
+++ b/src/d3d11/d3d11_texture.cpp
@@ -890,6 +890,12 @@ namespace dxvk {
       *ppvObject = ref(&m_interop);
       return S_OK;
     }
+
+    const GUID IID_IIWineD3D11Texture2D = {0x267dc993, 0xd15e, 0x4015, {0xaa, 0xac, 0xb7, 0x55, 0x9e, 0x22, 0x6c, 0xc3}};
+    if (riid == IID_IIWineD3D11Texture2D) {
+        /* wined3d interop interface, avoiding logging */
+        return E_NOINTERFACE;
+    }
     
     Logger::warn("D3D11Texture2D::QueryInterface: Unknown interface query");
     Logger::warn(str::format(riid));
