diff --git a/dlls/shell32/shlfileop.c b/dlls/shell32/shlfileop.c
index cd37594992..b7747301bd 100644
--- a/dlls/shell32/shlfileop.c
+++ b/dlls/shell32/shlfileop.c
@@ -826,6 +826,7 @@ int WINAPI SHCreateDirectoryExW(HWND hWnd, LPCWSTR path, LPSECURITY_ATTRIBUTES s
 
 	  if (ret && hWnd &&
 	      ret != ERROR_CANCELLED &&
+	      ret != ERROR_FILE_EXISTS &&
 	      ret != ERROR_ALREADY_EXISTS)
 	  {
 	    /* We failed and should show a dialog box */
