diff --git a/steam.cpp b/steam.cpp
--- a/steam.cpp
+++ b/steam.cpp
@@ -33,6 +33,7 @@
  * Windows version of Steam running. */

 #include <windows.h>
+#include <winternl.h>
 #include <shlobj.h>
 #include <string.h>
 #include <stdio.h>
@@ -52,7 +52,6 @@

 WINE_DEFAULT_DEBUG_CHANNEL(steam);

-EXTERN_C HANDLE CDECL __wine_make_process_system(void);

 #define ARRAY_SIZE(a) (sizeof(a) / sizeof(*a))

@@ -1177,7 +1177,7 @@ int main(int argc, char *argv[])
         setup_steam_registry();
         setup_steam_files();

-        wait_handle = __wine_make_process_system();
+        NtSetInformationProcess(GetCurrentProcess(), (PROCESS_INFORMATION_CLASS)1000 /*ProcessWineMakeProcessSystem*/, &wait_handle, sizeof(HANDLE));
         game_process = TRUE;
     }

