From 06dbd32747a1451deb12d1bebc9cbf5a56dfdca5 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 18 Mar 2021 16:54:44 -0400
Subject: [PATCH] mfplat: Stub out MFCreateDXGIDeviceManager, to avoid the d3d
 path.

---
 dlls/mfplat/main.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index 7fd713261fc..9e458ea49bb 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -8891,6 +8891,8 @@ HRESULT WINAPI MFCreateDXGIDeviceManager(UINT *token, IMFDXGIDeviceManager **man
 
     TRACE("%p, %p.\n", token, manager);
 
+    return E_NOTIMPL;
+
     if (!token || !manager)
         return E_POINTER;
 
-- 
2.30.2

