From 65baa873184632c22aad43ead553bba259f3d7c2 Mon Sep 17 00:00:00 2001
From: Robin Kertels <robin.kertels@gmail.com>
Date: Sat, 14 Aug 2021 18:07:54 +0200
Subject: [PATCH] vkd3d: Only disable raster based on SO stream if SO is used.

Signed-off-by: Robin Kertels <robin.kertels@gmail.com>
---
 libs/vkd3d/state.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libs/vkd3d/state.c b/libs/vkd3d/state.c
index 56daffb5..d8d275f1 100644
--- a/libs/vkd3d/state.c
+++ b/libs/vkd3d/state.c
@@ -3417,7 +3417,7 @@ static HRESULT d3d12_pipeline_state_init_graphics(struct d3d12_pipeline_state *s
     rs_desc_from_d3d12(&graphics->rs_desc, &desc->rasterizer_state);
     have_attachment = graphics->rt_count || graphics->dsv_format || is_dsv_format_unknown;
     if ((!have_attachment && !(desc->ps.pShaderBytecode && desc->ps.BytecodeLength))
-            || so_desc->RasterizedStream == D3D12_SO_NO_RASTERIZED_STREAM)
+            || (so_desc->NumEntries != 0 && so_desc->RasterizedStream == D3D12_SO_NO_RASTERIZED_STREAM))
         graphics->rs_desc.rasterizerDiscardEnable = VK_TRUE;
 
     rs_stream_info_from_d3d12(&graphics->rs_stream_info, &graphics->rs_desc, so_desc, vk_info);
