From 9b1b7ce57fcf42315ace4d1251a8ced53b0d1334 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 2 Apr 2020 15:39:16 -0500
Subject: [PATCH 31/36] mf: Make samplegrabber's media type handler more
 restrictive.

---
 dlls/mf/samplegrabber.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/dlls/mf/samplegrabber.c b/dlls/mf/samplegrabber.c
index 205298501d..cad3db033b 100644
--- a/dlls/mf/samplegrabber.c
+++ b/dlls/mf/samplegrabber.c
@@ -627,7 +627,7 @@ static HRESULT WINAPI sample_grabber_stream_type_handler_GetMediaTypeCount(IMFMe
     if (!count)
         return E_POINTER;
 
-    *count = 0;
+    *count = 1;
 
     return S_OK;
 }
@@ -635,12 +635,20 @@ static HRESULT WINAPI sample_grabber_stream_type_handler_GetMediaTypeCount(IMFMe
 static HRESULT WINAPI sample_grabber_stream_type_handler_GetMediaTypeByIndex(IMFMediaTypeHandler *iface, DWORD index,
         IMFMediaType **media_type)
 {
+    struct sample_grabber_stream *stream = impl_from_IMFMediaTypeHandler(iface);
+
     TRACE("%p, %u, %p.\n", iface, index, media_type);
 
     if (!media_type)
         return E_POINTER;
 
-    return MF_E_NO_MORE_TYPES;
+    if (index != 0)
+        return MF_E_NO_MORE_TYPES;
+
+    *media_type = stream->sink->media_type;
+    IMFMediaType_AddRef(stream->sink->media_type);
+
+    return S_OK;
 }
 
 static HRESULT WINAPI sample_grabber_stream_type_handler_SetCurrentMediaType(IMFMediaTypeHandler *iface,
@@ -656,6 +664,9 @@ static HRESULT WINAPI sample_grabber_stream_type_handler_SetCurrentMediaType(IMF
     if (!stream->sink)
         return MF_E_STREAMSINK_REMOVED;
 
+    if (FAILED(sample_grabber_stream_type_handler_IsMediaTypeSupported(iface, media_type, NULL)))
+        return MF_E_INVALIDMEDIATYPE;
+
     IMFMediaType_Release(stream->sink->media_type);
     stream->sink->media_type = media_type;
     IMFMediaType_AddRef(stream->sink->media_type);
-- 
2.25.1

