From 97bb40c902fc6638273b472c3edb975ff09fad32 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 2 Apr 2020 15:40:28 -0500
Subject: [PATCH 32/36] Use cloned topology for METopologySet event.

---
 dlls/mf/session.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index debb58de4d..c3e9f59635 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -1172,6 +1172,8 @@ static HRESULT session_set_current_topology(struct media_session *session, IMFTo
         return hr;
     }
 
+    session_raise_topology_set(session, session->presentation.current_topology, hr);
+
     session_collect_nodes(session);
 
     LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
@@ -1281,8 +1283,6 @@ static void session_set_topology(struct media_session *session, DWORD flags, IMF
         session_clear_presentation(session);
     }
 
-    session_raise_topology_set(session, topology, hr);
-
     /* With no current topology set it right away, otherwise queue. */
     if (topology)
     {
-- 
2.25.1

