From 4ecb6fe9c1c474dc1c1544bf493fffe1b491fdfd Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Thu, 2 Apr 2020 15:43:52 -0500
Subject: [PATCH 34/40] mf: Miscelaneous fixes to topology resolution.

---
 dlls/mf/topology.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/dlls/mf/topology.c b/dlls/mf/topology.c
index 7a4c95bea0..ce706a9ff7 100644
--- a/dlls/mf/topology.c
+++ b/dlls/mf/topology.c
@@ -2006,7 +2006,7 @@ static HRESULT topology_loader_resolve_branch(IMFTopologyNode *src, IMFMediaType
                         IMFTopologyNode_ConnectOutput(src, 0, node_dec, 0);
                         IMFTopologyNode_ConnectOutput(node_dec, 0, sink, 0);
 
-                        IMFActivate_ShutdownObject(activates_convs[i]);
+                        IMFActivate_ShutdownObject(activates_decs[i]);
                         return S_OK;
                     }
 
@@ -2145,12 +2145,14 @@ static HRESULT topology_loader_resolve_partial_topology(struct topology_node *sr
     }
     else
     {
-        if (SUCCEEDED(topology_loader_resolve_branch(clone_src, src_mediatypes[0], clone_sink, MF_CONNECT_DIRECT)))
-        {
-            topology_loader_add_branch(*full_topology, clone_src, clone_sink);
-            heap_free(src_mediatypes);
-            return S_OK;
-        }
+        for (method = MF_CONNECT_DIRECT; method < MF_CONNECT_RESOLVE_INDEPENDENT_OUTPUTTYPES; method++)
+            if (SUCCEEDED(topology_loader_resolve_branch(clone_src, src_mediatypes[0], clone_sink, method)))
+            {
+                TRACE("Successfully connected nodes with method %u\n", method);
+                topology_loader_add_branch(*full_topology, clone_src, clone_sink);
+                heap_free(src_mediatypes);
+                return S_OK;
+            }
     }
 
     heap_free(src_mediatypes);
-- 
2.25.1

