From 965d7dd8ff173eacb663c3dc1d23c3c55d7eb8eb Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 3 Apr 2020 13:09:11 -0500
Subject: [PATCH 38/40] Always request another sammple.

---
 dlls/mf/samplegrabber.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/mf/samplegrabber.c b/dlls/mf/samplegrabber.c
index cad3db033b..e655d72b3b 100644
--- a/dlls/mf/samplegrabber.c
+++ b/dlls/mf/samplegrabber.c
@@ -785,12 +785,13 @@ static HRESULT WINAPI sample_grabber_stream_timer_callback_Invoke(IMFAsyncCallba
                     if (FAILED(hr = sample_grabber_report_sample(stream->sink, item->u.sample, &sample_delivered)))
                         WARN("Failed to report a sample, hr %#x.\n", hr);
                     stream_release_pending_item(item);
+                    if (sample_delivered)
+                        sample_grabber_stream_request_sample(stream);
                     item = stream_get_next_item(stream);
                     if (item && item->type == ITEM_TYPE_SAMPLE)
                     {
                         if (FAILED(hr = stream_schedule_sample(stream, item)))
                             WARN("Failed to schedule a sample, hr %#x.\n", hr);
-                        sample_grabber_stream_request_sample(stream);
                         item = NULL;
                     }
                     break;
-- 
2.25.1

