From 4b86bfba104d2e3654da951ee09bece12699f40f Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 3 Apr 2020 13:09:30 -0500
Subject: [PATCH 39/40] Set BPS on AAC.

---
 dlls/winegstreamer/mfplat.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winegstreamer/mfplat.c b/dlls/winegstreamer/mfplat.c
index 6a0ff2d6e1..788c729261 100644
--- a/dlls/winegstreamer/mfplat.c
+++ b/dlls/winegstreamer/mfplat.c
@@ -965,6 +965,7 @@ static IMFMediaType* transform_to_media_type(GstCaps *caps)
                     struct aac_user_data *user_data = NULL;
 
                     IMFMediaType_SetGUID(media_type, &MF_MT_SUBTYPE, &MFAudioFormat_AAC);
+                    IMFMediaType_SetUINT32(media_type, &MF_MT_AUDIO_BITS_PER_SAMPLE, 16);
 
                     codec_data = gst_structure_get_value(info, "codec_data");
                     if (codec_data)
-- 
2.25.1

