From 9fa5509fbe50e1d6dc210218c163dc077f382063 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Fri, 3 Apr 2020 13:09:45 -0500
Subject: [PATCH 40/40] We probably shouldn't queue too many buffers.

---
 dlls/winegstreamer/media_source.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winegstreamer/media_source.c b/dlls/winegstreamer/media_source.c
index 171b8a7d7e..d2d02be676 100644
--- a/dlls/winegstreamer/media_source.c
+++ b/dlls/winegstreamer/media_source.c
@@ -448,6 +448,7 @@ static HRESULT media_stream_constructor(struct media_source *source, GstPad *pad
 
     g_object_set(This->appsink, "emit-signals", TRUE, NULL);
     g_object_set(This->appsink, "sync", FALSE, NULL);
+    g_object_set(This->appsink, "max-buffers", 10, NULL);
     g_signal_connect(This->appsink, "new-sample", G_CALLBACK(stream_new_sample_wrapper), This);
     g_signal_connect(This->appsink, "eos", G_CALLBACK(stream_eos_wrapper), This);
 
-- 
2.25.1

