diff --git a/src/util/config/config.cpp b/src/util/config/config.cpp
index da46e3eb0..8f0fe528f 100644
--- a/src/util/config/config.cpp
+++ b/src/util/config/config.cpp
@@ -251,6 +251,10 @@ namespace dxvk {
       { "d3d9.memoryTrackTest",             "True" },
       { "d3d9.maxAvailableMemory",          "1073741824" },
     }} },
+    /* Warlock: Master of the Arcane        */
+    { R"(\\Game\.exe$)", {{
+      { "d3d9.deviceLocalMemory",        "True" },
+    }} },
   }};
 
 
diff --git a/src/d3d9/d3d9_options.h b/src/d3d9/d3d9_options.h
index 3f1414941..47cf8043b 100644
--- a/src/d3d9/d3d9_options.h
+++ b/src/d3d9/d3d9_options.h
@@ -104,6 +104,9 @@ namespace dxvk {
     /// Support VCACHE query
     bool supportVCache;
 
+    /// Allow DEVICE_LOCAL memory for shader constant buffers
+    bool deviceLocalMemory;
+
     /// Forced aspect ratio, disable other modes
     std::string forceAspectRatio;
   };
diff --git a/src/d3d9/d3d9_options.cpp b/src/d3d9/d3d9_options.cpp
index 279952588..5f557217d 100644
--- a/src/d3d9/d3d9_options.cpp
+++ b/src/d3d9/d3d9_options.cpp
@@ -59,7 +59,7 @@ namespace dxvk {
     this->invariantPosition     = config.getOption<bool>    ("d3d9.invariantPosition",     false);
     this->memoryTrackTest       = config.getOption<bool>    ("d3d9.memoryTrackTest",       false);
     this->supportVCache         = config.getOption<bool>    ("d3d9.supportVCache",         vendorId == 0x10de);
-
+    this->deviceLocalMemory     = config.getOption<bool>    ("d3d9.deviceLocalMemory",     false);
     this->forceAspectRatio      = config.getOption<std::string>("d3d9.forceAspectRatio",   "");
 
     // If we are not Nvidia, enable general hazards.
@@ -71,4 +71,4 @@ namespace dxvk {
     applyTristate(this->d3d9FloatEmulation, config.getOption<Tristate>("d3d9.floatEmulation", Tristate::Auto));
   }
 
-}
\ No newline at end of file
+}
diff --git a/src/d3d9/d3d9_device.cpp b/src/d3d9/d3d9_device.cpp
index 9ff65aeb0..a6c3dd5fd 100644
--- a/src/d3d9/d3d9_device.cpp
+++ b/src/d3d9/d3d9_device.cpp
@@ -4614,6 +4614,13 @@ namespace dxvk {
     VkMemoryPropertyFlags memoryFlags = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
                                       | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
 
+    if(m_d3d9Options.deviceLocalMemory)
+    {
+        memoryFlags = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT
+                    | VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
+                    | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT;
+    }
+
     info.stages = VK_PIPELINE_STAGE_VERTEX_SHADER_BIT;
     info.size   = m_vsLayout.totalSize();
     m_consts[DxsoProgramTypes::VertexShader].buffer = m_dxvkDevice->createBuffer(info, memoryFlags);
