diff --git a/dlls/ntdll/version.c b/dlls/ntdll/version.c
index 61e48f6fbd..cc2b690b61 100644
--- a/dlls/ntdll/version.c
+++ b/dlls/ntdll/version.c
@@ -482,6 +482,7 @@ void version_init(void)
     BOOL got_win_ver = FALSE;
     const WCHAR *p, *appname = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
     WCHAR appversion[MAX_PATH+20];
+    SYSTEM_BASIC_INFORMATION sbi;
 
     current_version = &VersionData[WIN7];
 
@@ -547,6 +548,9 @@ done:
     user_shared_data->NtMinorVersion     = current_version->dwMinorVersion;
     user_shared_data->SuiteMask          = current_version->wSuiteMask;
 
+    virtual_get_system_info(&sbi);
+    user_shared_data->NumberOfPhysicalPages = sbi.MmNumberOfPhysicalPages;
+
     TRACE( "got %d.%d platform %d build %x name %s service pack %d.%d product %d\n",
            current_version->dwMajorVersion, current_version->dwMinorVersion,
            current_version->dwPlatformId, current_version->dwBuildNumber,
