@@ -, +, @@ 
---
 dlls/ntdll/signal_x86_64.c | 1 +
 1 file changed, 1 insertion(+)
--- a/dlls/ntdll/signal_x86_64.c	
+++ a/dlls/ntdll/signal_x86_64.c	
@@ -4204,6 +4204,7 @@ static void init_thread_context( CONTEXT *context, LPTHREAD_START_ROUTINE entry,
 {
     __asm__( "movw %%cs,%0" : "=m" (context->SegCs) );
     __asm__( "movw %%ss,%0" : "=m" (context->SegSs) );
+    __asm__( "movw %0,%%ds" :: "m" (context->SegSs) );
     context->Rcx    = (ULONG_PTR)entry;
     context->Rdx    = (ULONG_PTR)arg;
     context->Rsp    = (ULONG_PTR)NtCurrentTeb()->Tib.StackBase - 0x28;
-- 
