diff --git a/dlls/msvcrt/scanf.h b/dlls/msvcrt/scanf.h
index de2a2af975..9f521f53bd 100644
--- a/dlls/msvcrt/scanf.h
+++ b/dlls/msvcrt/scanf.h
@@ -302,7 +302,23 @@ _FUNCTION_ {
 			seendigit=1;
                     }
 		    /* okay, done! */
-		    if (!seendigit) break; /* not a valid number */
+		    if (!seendigit)
+                    {
+                        if (nch!=_EOF_ && (nch=='n' || nch=='N'))
+                        {
+                            nch = _GETC_(file);
+                            if (nch!=_EOF_ && (nch=='a' || nch=='A'))
+                            {
+                                nch = _GETC_(file);
+                                if (nch!=_EOF_ && (nch=='n' || nch=='N'))
+                                {
+                                    cur=0; /* Superfluous? */
+                                    seendigit=1;
+                                }
+                            }
+                        }
+		    	if (!seendigit) break; /* not a valid number */
+                    }
                     st = 1;
                     if (!suppress) {
 #define _SET_NUMBER_(type) *va_arg(ap, type*) = negative ? -cur : cur
@@ -341,7 +357,22 @@ _FUNCTION_ {
 
                     /* get first digit. */
                     if (*locinfo->lconv->decimal_point != nch) {
-                        if (!_ISDIGIT_(nch)) break;
+                        if (!_ISDIGIT_(nch))
+                        {
+                            if (nch!=_EOF_ && (nch=='n' || nch=='N'))
+                            {
+                                nch = _GETC_(file);
+                                if (nch!=_EOF_ && (nch=='a' || nch=='A'))
+                                {
+                                    nch = _GETC_(file);
+                                    if (nch!=_EOF_ && (nch=='n' || nch=='N'))
+                                    {
+                                        nch='0';
+                                    }
+                                }
+                            }
+                            if (!_ISDIGIT_(nch)) break; /* not a valid number */
+                        }
                         d = nch - '0';
                         nch = _GETC_(file);
                         if (width>0) width--;
 
