From bad26eab9afd475b69fe3a9320583f90722216fb Mon Sep 17 00:00:00 2001
From: Paul Gofman <gofmanp@gmail.com>
Date: Tue, 10 Mar 2020 20:01:51 +0300
Subject: [PATCH] ntdll: Tweak heap arena size to help some use after free
 scenarios.

---
 dlls/ntdll/heap.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/heap.c b/dlls/ntdll/heap.c
index d7232b47d6..5ca7ad8e21 100644
--- a/dlls/ntdll/heap.c
+++ b/dlls/ntdll/heap.c
@@ -56,6 +56,7 @@ typedef struct tagARENA_INUSE
     DWORD  size;                    /* Block size; must be the first field */
     DWORD  magic : 24;              /* Magic number */
     DWORD  unused_bytes : 8;        /* Number of bytes in the block not used by user data (max value is HEAP_MIN_DATA_SIZE+HEAP_MIN_SHRINK_SIZE) */
+    char   reserved[16];
 } ARENA_INUSE;
 
 typedef struct tagARENA_FREE
@@ -91,7 +92,7 @@ typedef struct
 #define ARENA_FREE_FILLER      0xfeeefeee
 
 /* everything is aligned on 8 byte boundaries (16 for Win64) */
-#define ALIGNMENT              (2*sizeof(void*))
+#define ALIGNMENT              (32)
 #define LARGE_ALIGNMENT        16  /* large blocks have stricter alignment */
 #define ARENA_OFFSET           (ALIGNMENT - sizeof(ARENA_INUSE))
 
-- 
2.24.1

