diff --git a/dlls/ntdll/heap.c b/dlls/ntdll/heap.c
index a553e4a0a6..ed52d8d5b3 100644
--- a/dlls/ntdll/heap.c
+++ b/dlls/ntdll/heap.c
@@ -56,8 +56,15 @@ typedef struct tagARENA_INUSE
     DWORD  size;                    /* Block size; must be the first field */
     DWORD  magic : 24;              /* Magic number */
     DWORD  unused_bytes : 8;        /* Number of bytes in the block not used by user data (max value is HEAP_MIN_DATA_SIZE+HEAP_MIN_SHRINK_SIZE) */
+#ifdef _WIN64
+    char reserved[8];
+#endif
 } ARENA_INUSE;
 
+#ifdef _WIN64
+C_ASSERT( sizeof(ARENA_INUSE) == 16 );
+#endif
+
 typedef struct tagARENA_FREE
 {
     DWORD                 size;     /* Block size; must be the first field */
-- 
2.24.1

