From a90109abaa6114c785786b0b41dcaaa103ab59dc Mon Sep 17 00:00:00 2001
From: Anna Lasky <alasky@codeweavers.com>
Date: Fri, 5 Feb 2021 08:31:07 -0600
Subject: [PATCH] wine.inf: Use built-in atiadlxx for Need For Speed

---
 loader/wine.inf.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 928c6883372..fd529938508 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -4030,6 +4030,7 @@ HKCU,Software\Wine\AppDefaults\h1_sp64_ship.exe\DllOverrides,"atiadlxx",,"builti
 HKCU,Software\Wine\AppDefaults\h1_mp64_ship.exe\DllOverrides,"atiadlxx",,"builtin"
 HKCU,Software\Wine\AppDefaults\iw7_ship.exe\DllOverrides,"atiadlxx",,"builtin"
 HKCU,Software\Wine\AppDefaults\BlackOps3.exe\DllOverrides,"atiadlxx",,"builtin"
+HKCU,Software\Wine\AppDefaults\NFS16.exe\DllOverrides,"atiadlxx",,"builtin"
 HKCU,Software\Wine\AppDefaults\ShadowOfWar.exe\DllOverrides,"atiadlxx",,"builtin"
 HKCU,Software\Wine\AppDefaults\ShadowOfWar.exe\DllOverrides,"amd_ags_x64",,""
 ;;Likely want *80 and *90 too, but those require removing Wine's manifest files.
