diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 3d757d9a2f8..799234e1837 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -6047,6 +6047,7 @@ HKCU,Software\Wine\AppDefaults\WRC10.exe\DllOverrides,"xinput1_3",,"native"
 HKCU,Software\Wine\AppDefaults\Paradox Launcher.exe\DllOverrides,"libglesv2",,"disabled"
 HKCU,Software\Wine\AppDefaults\Nickelodeon All-Star Brawl.exe\DllOverrides,"winusb",,"disabled"
 HKCU,Software\Wine\AppDefaults\nw.exe\DllOverrides,"libglesv2",,"disabled"
+HKCU,Software\Wine\AppDefaults\gotg.exe\DllOverrides,"atiadlxx",,"builtin"
 ;;Likely want *80 and *90 too, but those require removing Wine's manifest files.
 HKCU,Software\Wine\DllOverrides,"atl100",0x2,"native,builtin"
 HKCU,Software\Wine\DllOverrides,"msvcp100",0x2,"native,builtin"
