From 8dcfe3073f04551eb3ab51b162ba750741fd7d53 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 2 Apr 2020 20:41:43 +0200
Subject: [PATCH] ntdll: Guard seccomp with WINESECCOMP env var.

---
 dlls/ntdll/unix/signal_x86_64.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index 97e75ff47e..97bc385e43 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -1034,6 +1034,16 @@ static int sc_seccomp(unsigned int operation, unsigned int flags, void *args)
 
 static void install_bpf(struct sigaction *sig_act)
 {
+    static int enable_seccomp = -1;
+
+    if (enable_seccomp == -1)
+        enable_seccomp = getenv("WINESECCOMP") && atoi(getenv("WINESECCOMP"));
+
+    if (!enable_seccomp)
+        return;
+
+    MESSAGE("wine: enabling seccomp syscall filters.\n");
+    
 #ifdef HAVE_SECCOMP
 #   ifndef SECCOMP_FILTER_FLAG_SPEC_ALLOW
 #       define SECCOMP_FILTER_FLAG_SPEC_ALLOW (1UL << 2)
