From c2351cd9b44910a9be03f0c84e7fbb992a783adf Mon Sep 17 00:00:00 2001
From: Roberto Pungartnik <rpungartnik@gmail.com>
Date: Sat, 29 May 2021 12:43:54 -0300
Subject: [PATCH] ntdll: Fix IOCTL_AFD_WINE_SENDMSG irda member misreference.

Signed-off-by: Roberto Augusto Pungartnik <rpungartnik@gmail.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/ntdll/unix/socket.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/unix/socket.c b/dlls/ntdll/unix/socket.c
index 6112a1260cb..db65752a031 100644
--- a/dlls/ntdll/unix/socket.c
+++ b/dlls/ntdll/unix/socket.c
@@ -255,10 +255,10 @@ static socklen_t sockaddr_to_unix( const struct WS_sockaddr *wsaddr, int wsaddrl
         memcpy( &win, wsaddr, sizeof(win) );
         uaddr->irda.sir_family = AF_IRDA;
         if (sscanf( win.irdaServiceName, "LSAP-SEL%u", &lsap_sel ) == 1)
-            uaddr->sir_lsap_sel = lsap_sel;
+            uaddr->irda.sir_lsap_sel = lsap_sel;
         else
         {
-            uaddr->sir_lsap_sel = LSAP_ANY;
+            uaddr->irda.sir_lsap_sel = LSAP_ANY;
             memcpy( uaddr->irda.sir_name, win.irdaServiceName, sizeof(win.irdaServiceName) );
         }
         memcpy( &uaddr->irda.sir_addr, win.irdaDeviceID, sizeof(win.irdaDeviceID) );
