diff --git a/dlls/evr/mixer.c b/dlls/evr/mixer.c
index 187a13be364..354f89341c9 100644
--- a/dlls/evr/mixer.c
+++ b/dlls/evr/mixer.c
@@ -814,7 +814,8 @@ static HRESULT WINAPI video_mixer_transform_SetInputType(IMFTransform *iface, DW
 
 static HRESULT WINAPI video_mixer_transform_SetOutputType(IMFTransform *iface, DWORD id, IMFMediaType *type, DWORD flags)
 {
-    const unsigned int equality_flags = MF_MEDIATYPE_EQUAL_MAJOR_TYPES | MF_MEDIATYPE_EQUAL_FORMAT_TYPES;
+    const unsigned int equality_flags = MF_MEDIATYPE_EQUAL_MAJOR_TYPES |
+            MF_MEDIATYPE_EQUAL_FORMAT_TYPES | MF_MEDIATYPE_EQUAL_FORMAT_DATA;
     struct video_mixer *mixer = impl_from_IMFTransform(iface);
     HRESULT hr = MF_E_INVALIDMEDIATYPE;
     unsigned int i, compare_flags;
