From 501b068024375386484c6a8616a406b21ab15406 Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Thu, 30 Jan 2020 14:44:00 -0600
Subject: [PATCH] user32: HACK: Don't resize new windows to fit the display

---
 dlls/user32/win.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/user32/win.c b/dlls/user32/win.c
index fa593a4a29a..803259b351f 100644
--- a/dlls/user32/win.c
+++ b/dlls/user32/win.c
@@ -1607,12 +1607,17 @@ HWND WIN_CreateWindowEx( CREATESTRUCTW *cs, LPCWSTR className, HINSTANCE module,
 
     cx = cs->cx;
     cy = cs->cy;
+#if 0
+    /* HACK: This code changes the window's size to fit the display. However,
+     * some games (Bayonetta, Dragon's Dogma) will then have the incorrect
+     * render size. So just let windows be too big to fit the display. */
     if ((cs->style & WS_THICKFRAME) || !(cs->style & (WS_POPUP | WS_CHILD)))
     {
         MINMAXINFO info = WINPOS_GetMinMaxInfo( hwnd );
         cx = max( min( cx, info.ptMaxTrackSize.x ), info.ptMinTrackSize.x );
         cy = max( min( cy, info.ptMaxTrackSize.y ), info.ptMinTrackSize.y );
     }
+#endif
 
     if (cx < 0) cx = 0;
     if (cy < 0) cy = 0;
